File: app/src/main/java/com/example/mhdstuff/activity/DeparturePostDetailActivity.java
package com.example.mhdstuff.activity;

import android.content.Context;
import android.os.Bundle;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;

import com.example.mhdstuff.DepartureEntryItemAdapter;
import com.example.mhdstuff.R;
import com.example.mhdstuff.activity.data.DelaysDataHolder;
import com.example.mhdstuff.activity.listview.AbstractListViewActivity;
import com.example.mhdstuff.exception.AppException;
import com.example.mhdstuff.exception.RequestException;
import com.example.mhdstuff.parsing.storage.IdStorage;
import com.example.mhdstuff.parsing.types.Post;
import com.example.mhdstuff.parsing.types.Time;
import com.example.mhdstuff.parsing.types.departure.Departure;
import com.example.mhdstuff.util.OfflineDepartures;
import com.example.mhdstuff.util.request.RequestHelper;
import com.google.gson.JsonObject;

import java.util.List;

public class DeparturePostDetailActivity extends AbstractListViewActivity {


    private final JsonObject delays;
    private Post post;

    public DeparturePostDetailActivity() {
        super("", R.layout.activity_deparute_post_detail, R.id.departure_content);

        this.delays = DelaysDataHolder.getDelays();
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        post = getIntent().getParcelableExtra("post");
        setName(post.name());
    }

    @Override
    protected RecyclerView.Adapter<?> getAdapter(Context context, IdStorage storage) {
        JsonObject stopDelays = new JsonObject();
        try {
            stopDelays = RequestHelper.getStopDelays(post.stopID());
        } catch (RequestException e) {
            runOnUiThread(() -> e.showError(this, AppException.NotificationType.SNACK_BAR));
        }

        List<Departure> departureList = OfflineDepartures.getOfflineForPost(storage, post.stopID(),post.postID(), -1, Time.ZERO, delays);

        Departure departure = departureList.stream().filter(dep -> dep.postID() == post.postID()).findFirst().orElse(null);

        TextView title = findViewById(R.id.departure_title);
        title.setText(departure.name());

        var adapter = new DepartureEntryItemAdapter(this, departure.entries(),storage.apiStorage(), stopDelays);

        Time now = Time.now();

        int firstPos = -1;
        for (int i = 0; i < departure.entries().size(); i++) {
            var entry = departure.entries().get(i);
            if (!entry.timeMark().time().isBefore(now)) {
                firstPos = i;
                break;
            }
        }
        int finalFirstPos = firstPos;
        runOnUiThread(() -> {
            RecyclerView recyclerView = findViewById(recycleViewId);
            recyclerView.scrollToPosition(finalFirstPos);
        });

        return adapter;
    }
}

File: app/src/main/java/me/miran/libreinfo/activity/DeparturePostDetailActivity.kt
package me.miran.libreinfo.activity

import androidx.compose.animation.Crossfade
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.colorResource
import androidx.compose.ui.unit.dp
import com.google.gson.JsonObject
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import me.miran.libreinfo.R
import me.miran.libreinfo.activity.base.KBaseActivity
import me.miran.libreinfo.activity.data.DelaysDataHolder
import me.miran.libreinfo.exception.RequestException
import me.miran.libreinfo.parsing.storage.manager.AppContainer
import me.miran.libreinfo.parsing.storage.manager.IdStorage
import me.miran.libreinfo.parsing.types.Post
import me.miran.libreinfo.parsing.types.Time
import me.miran.libreinfo.parsing.types.departure.Departure
import me.miran.libreinfo.util.DeparturesSettings
import me.miran.libreinfo.util.LocalDeparturesSettings
import me.miran.libreinfo.util.OfflineDepartures
import me.miran.libreinfo.util.Text
import me.miran.libreinfo.util.request.RequestHelper

class DeparturePostDetailActivity : KBaseActivity("") {

    private var departuresSettings by mutableStateOf(DeparturesSettings())

    override fun onResume() {
        super.onResume()
        departuresSettings = DeparturesSettings.fromPrefs()
    }

    @Composable
    override fun CreateElements() {
        val post = intent.getParcelableExtra<Post>("post")!!


        LaunchedEffect(post) { name = Text.literal(post.name) }

        val context = LocalContext.current

        var storage: IdStorage? by remember { mutableStateOf(null) }
        var departureResult: Departure? by remember { mutableStateOf(null) }
        var stopDelays by remember { mutableStateOf(JsonObject()) }

        val delays = DelaysDataHolder.getDelays()
        LaunchedEffect(Unit) {
            stopDelays = withContext(Dispatchers.IO) {
                try {
                    RequestHelper.getStopDelays(context, post.stop.id)
                } catch (e: RequestException) {
                    showErrorSnackBar(e)
                    JsonObject()
                }
            }

            val (depsRes, storageRes) = withContext(Dispatchers.IO) {
                val storage = AppContainer.storageProvider.getInstance()

                val departureList = OfflineDepartures.getOfflineForPost(
                    storage,
                    post.stop.id.internal,
                    post.postID,
                    -1,
                    Time.ZERO,
                    delays
                )

                val res = departureList.stream().filter { dep: Departure? -> dep!!.postID == post.postID }
                        .findFirst().orElse(null)

                Pair(res, storage)
            }

            departureResult = depsRes
            storage = storageRes
        }

        Crossfade(targetState = departureResult) { departure ->
            if (departure != null && storage != null) {
                CompositionLocalProvider(LocalDeparturesSettings provides departuresSettings) {
                    DepartureDetail(departure, storage!!.apiStorage, stopDelays)
                }
            } else {
                DepartureDetailShimmer(post)
            }
        }
    }


    @Composable
    fun DepartureDetailShimmer(post: Post) {
        val shimmer = rememberActivityShimmer()
        val color = colorResource(R.color.widget_background)
        Container(
            innerPadding = 0.dp,
            modifier = Modifier.padding(horizontal = 16.dp, vertical = 8.dp)
        ) {
            LazyColumn(Modifier.padding(vertical = 8.dp, horizontal = 6.dp)) {
                stickyHeader {
                    DeparturePostHeader(
                        post.name, Modifier
                            .background(color)
                            .clickable(interactionSource = null, indication = null) {})

                }

                items(30) { _ ->
                    DepartureEntryRowShimmer(shimmer)
                }
            }
        }
    }

}